libsolutil/CommonIO.cpp

```diff
@@ -78,6 +78,7 @@ string solidity::util::readStandardInput()
 		ret.append(tmp);
 		ret.append("\n");
 	}
+	ret.pop_back();
```

Comment:
The function still replaces line endings with `\n` so I think it does not fix the issue on Windows. Unfortunately there are no tests for it so CI does not fail.

I think we really need at least a simple test here. It should check that its output is the same as from `readFileAsString()`.

It would be hard to test with hard-coded `cin` but I already have that changed to a stream parameter in #11520. I'll extract this part into a separate PR so that you can rebase on it.

Reply:
Done. @TerranCivilian can you rebase your PR on #11589 and change the base branch on github to `testable-common-io`? Then you'll have everything you need to add a new test case in `test/libsolutil/CommonIO.cpp`.

Reply:
Sure, I can do this after work, i.e. in about 8 hours

Reply:
Thanks. By the way, there's no hurry, just wanted to make sure you have everything ready once you have time to get back to it :)

Reply:
@cameel : Hi, I tried adding a test here:
https://github.com/TerranCivilian/solidity/commit/2b8bde907f50c08c8c8e03ad0fc1e2b6cb00ea81

Is this similar to what you had in mind? This test demonstrates the newline bug.

Reply:
Yeah. That's exactly it. I'd add a few more cases because currently the function is not covered by any at all but that's the right direction.

You can also shorten your case to this:
```c++
BOOST_TEST(readUntilEnd(istringstream("ABC\ndef")) == "ABC\ndef");
```

Reply:
@cameel Ok I'll add a few more tests. As for actually _fixing_ the issue, what do you think? Above you mentioned that it could read content in binary mode like `readFileAsString()`. I will look into this. After a little digging, I might use `istreambuf_iterator`, but this could be the wrong direction :-)

Regarding your suggestion to shorten the test case: I saw that and was like oh duh, of course I should do it like that. But the compiler is complaining about using an rvalue as `readUntilEnd` argument I think:
`error: cannot bind non-const lvalue reference of type ‘std::istream&’ {aka ‘std::basic_istream<char>&’} to an rvalue of type ‘std::basic_istream<char>`

This is part of C++ and other programming languages that I don't fully understand yet- the difference between using a variable name vs. the function call itself for a value. Apparently, it doesn't like using the function call as the value here; I must assign it to a variable name first :shrug:

Reply:
> After a little digging, I might use `istreambuf_iterator`, but this could be the wrong direction :-)

The direction is right but I think that `istreambuf_iterator` will process the file character by character which could add significant overhead. I think it will be more efficient to just get the content from [`rdbuf()`](https://en.cppreference.com/w/cpp/io/basic_ios/rdbuf) and copy it into a string.

> But the compiler is complaining about using an rvalue as `readUntilEnd` argument I think:

Ah, true. My bad. The compiler is right. It would work fine with a `const&` argument but the function expects a non-`const` reference to an object instead - can't be `const` because reading from a stream necessarily changes its internal state. The compiler does not let us give it a temporary instance since it will be discarded and all the changes the function makes will be lost. We do not care about these changes but the compiler does not know that. So yeah, it unfortunately won't work - you have to use a variable.

> This is part of C++ and other programming languages that I don't fully understand yet- the difference between using a variable name vs. the function call itself for a value.

This is simply a constructor call. You can do that with any type and the result is a new instance of an object allocated on the stack. In C to do something like that you would instead allocate the variable yourself write a function that accepts an address and initializes it. The way it's done in C++ allows you to immediately assign that result to something or even pass into in a function call.